File: 22-Problem-B75-Duplicate-Integer/22-Problem-B75-Duplicate-Integer.js
// Given an integer array nums, return true if any value appears more than once in the array, otherwise return false


const nums1 = [0, 1, 2, 3] // no dups
const nums2 = [1, 2, 3, 3] // one dup
const nums3 = [1, 1, 2, 2, 3] // two dups
const nums4 = [1, 2, 3, 3, 3] // a triple dup
const nums5 = [] // edge empty array
const nums6 = [1] // edge single element array
const nums7 = [1, 'a', 3.3, "four!", 1] // not applicable because should be only nums

// use hashmap or set to keep track of seen nums as we iterate through array
// if we see a num that is already in the hashmap/set, return true
// if we iterate through entire array and don't see a dup, return false
// both are O(n) time complexity and O(n) space complexity

const containsDuplicate = (nums) => {
    // should we be checking for array of mixed type?
    const seenNums = new Set()
    for (let num of nums) {
        if (seenNums.has(num)) return true
        seenNums.add(num)
    }
    return false
}

// or

const hasDuplicate = (nums) => {
  // should we be checking for array of mixed type?
    const seenNums = {}
    for (let num of nums) {
        if (seenNums[num]) return true
        seenNums[num] = true
    }
    return false
}

console.log(containsDuplicate(nums1)) // false
console.log(containsDuplicate(nums2)) // true
console.log(containsDuplicate(nums3)) // true
console.log(containsDuplicate(nums4)) // true
console.log(containsDuplicate(nums5)) // false
console.log(containsDuplicate(nums6)) // false
console.log(containsDuplicate(nums7)) // true ***


console.log(hasDuplicate(nums1)) // false
console.log(hasDuplicate(nums2)) // true
console.log(hasDuplicate(nums3)) // true
console.log(hasDuplicate(nums4)) // true
console.log(hasDuplicate(nums5)) // false
console.log(hasDuplicate(nums6)) // false
console.log(hasDuplicate(nums7)) // true ***

File: 22-Problem-B75-Duplicate-Integer/22-Problem-B75-Duplicate-Integer.go

// arrays have fixed size in go - slices do not

package main

import (
	"fmt"
)

func containsDuplicate(nums []int) bool {
	// should we be checking for array of mixed type?
		var seenNums = make(map[int]bool);

		for _, num := range nums {
			if seenNums[num] {
				return true
			}
			seenNums[num] = true
		}


		return false
}


	// no sets in go lang so rather than implement set in hash just skip it



func main() {


	var nums1 = []int{0, 1, 2, 3};
	var nums2 = []int{1, 2, 3, 3} // one dup
	var nums3 = []int{1, 1, 2, 2, 3} // two dups
	var nums4 = []int{1, 2, 3, 3, 3} // a triple dup
	var nums5 = []int{} // edge empty array
	var nums6 = []int{1} // edge single element array
	// const nums7 = [4]int{1, 'a', 3.3, "four!", 1} // not applicable because should be only nums


	fmt.Println(containsDuplicate(nums1)) // false
	fmt.Println(containsDuplicate(nums2)) // true
	fmt.Println(containsDuplicate(nums3)) // true
	fmt.Println(containsDuplicate(nums4)) // true
	fmt.Println(containsDuplicate(nums5)) // false
	fmt.Println(containsDuplicate(nums6)) // false
	// console.log(containsDuplicate(nums7)) // true ***

}
